docs/etcd.md

```diff
@@ -0,0 +1,405 @@
+  - It is a database focused on consistency.
+  - The maximum recommended size of all data is [8GB](https://etcd.io/docs/v3.5/dev-guide/limit/).
+- Multiple keys from a `prefix/`, or `start-end` range can be loaded using [Range Queries](https://etcd.io/docs/v3.5/learning/api/#range).
+- The database provides `if, then, else` [transactions](https://etcd.io/docs/v3.5/learning/api/#transaction).
+  - This is a different concept than transactions in relational databases.
+- The [Lease API](https://etcd.io/docs/v3.5/learning/api/#lease-api) enables deletion of keys on client disconnection or outage.
+- The [Watch API](https://etcd.io/docs/v3.5/learning/api/#watch-api) is very important to us.
+  - The database sends information about key changes to clients.
+  - The client does not have to periodically query the database state.
+  - This allows the necessary part of the database state to be stored in memory.
+  - The in-memory state is updated after each change through the Watch API.
+  - However, it is necessary to consider that cluster nodes will receive the update at slightly different times.
+    - This must be accounted for in the system architecture.
+- The [Helm chart](https://artifacthub.io/packages/helm/bitnami/etcd) is used for deployment, so we are not dependent on any cloud provider.
+- etcd can be [embedded](https://pkg.go.dev/go.etcd.io/etcd/server/v3/embed) into our binaries:
+  - We don't use this now, except for a few tests, but it could be useful in the future.
+  - In other words, we can distribute the entire Stream API, all its components, and the database in one binary without external dependencies.
+
+## Framework - `etcdop`
+
+Since etcd is our main database in the Stream API, and we have complex business logic there, we needed a high-level framework on top of etcd.
+
+Unlike relational databases, there is no ORM-like framework for etcd:
+- The main reason is that it is a database for specific use in distributed systems.
+- It would be overkill for a simple API, e-shop, etc.
+- Kubernetes has helpers and libraries that make it easier to work with etcd.
+  - However, they are internal libraries whose interface can change at any time.
+  - Additionally, they don't use generic types (because they weren't available in Go at that time).
+  - Generic types can make the job a lot easier.
+
+**The `etcdop` framework provides:**
+- Saving entities in `JSON` format, as etcd itself only stores bytes - see [Serde](#serde).
+- `PrefixT[T]` and `KeyT[T]` generic types - abstractions for etcd prefix and key.
+  - `T` is the type of the entity, see [Schema](#schema).
+- A high-level wrapper for basic etcd operations, see [Basic Operations](#basic-operations).
+- Combination of basic operations into a transaction, see [TxnOp](#txnop).
+- Combination of basic operations and transactions into an atomic operation with multiple read phases, see [AtomicOp](#atomicop).
```

Comment:
Which of these 3 operations do I prefer to use.
1. Basic operations are similar as I would be using `etcdctl` or?
2. When do I use `TxnOp`? Only when I have intent to `merge` 2 transactions together?
3. When do I use `AtomicOp`? Do I use it only when PUT and GET operations should be done together? Could you provide some `complex` AtomicOp and TxnOp and go little bit in depth of the example? For example I want to create File only when Sink and Source exists etc. 
4. What was the flow that you come with such solution?